### PR TITLE
NH-3749 - Remove unnecessary comma in CREATE TABLE statement

### PR DESCRIPTION
**Pull request by lesnyh**
I develop NHibernate Driver for DBMS Linter SQL and found a problem in file Nhibernate\Mapping\Table.cs. The problem is in the following code:
foreach (UniqueKey uk in UniqueKeyIterator) 
{ 
    buf.Append(',').Append(uk.SqlConstraintString(dialect)); 
}
This code produces unnecessary comma in CREATE TABLE statement if dialect doesn't support not null unique constraints. For example, in test Nhibernate.Test.Component.Basic.ComponentWithUniqueConstraintTests.CanBePersistedWithUniqueValues we have the following query:
create table Employee (Id BIGINT not null, Name VARCHAR(255) null, Dob DATE null, HireDate DATE null, primary key (Id),); 
I suggest that we check dialect.SupportsNotNullUnique before running foreach loop. Something like this:
if (dialect.SupportsNotNullUnique)
{
    foreach (UniqueKey uk in UniqueKeyIterator)
    {
        buf.Append(',').Append(uk.SqlConstraintString(dialect));
    }
}

### CODE REVIEW

**Comment by hazzik** [comment]
@lesnyh can you add some tests?

**Comment by lesnyh** [comment]
Yes, I will try to add some tests. But to reproduce this problem we need DBMS that doesn't support not null unique constraints (dialect.SupportsNotNullUnique == false). I use DBMS Linter SQL Server and have problems in 21 tests:
NHibernate.Test.Component.Basic.ComponentWithUniqueConstraintTests.CanBePersistedWithUniqueValues
NHibernate.Test.Generatedkeys.Identity.SimpleIdentityGeneratedFixture.SequenceIdentityGenerator
NHibernate.Test.MappingByCode.IntegrationTests.NH3269.FixtureNonPublicProperty.ShouldNotThrowExceptionWhenTryingToSaveInherited2WithDuplicateName
NHibernate.Test.MappingByCode.IntegrationTests.NH3269.FixturePublicProperty.ShouldNotThrowExceptionWhenTryingToSaveInherited2WithDuplicateName
NHibernate.Test.Naturalid.Immutable.ImmutableNaturalIdFixture.NaturalIdCache
NHibernate.Test.Naturalid.Immutable.ImmutableNaturalIdFixture.NaturalIdCheck
NHibernate.Test.Naturalid.Immutable.ImmutableNaturalIdFixture.Update
NHibernate.Test.Naturalid.Mutable.MutableNaturalIdFixture.NaturalIdCache
NHibernate.Test.Naturalid.Mutable.MutableNaturalIdFixture.NonexistentNaturalIdCache
NHibernate.Test.Naturalid.Mutable.MutableNaturalIdFixture.Querying
NHibernate.Test.Naturalid.Mutable.MutableNaturalIdFixture.ReattachmentNaturalIdCheck
NHibernate.Test.NHSpecificTest.NH1989.Fixture.SecondLevelCacheWithDifferentRegionsFuture
NHibernate.Test.NHSpecificTest.NH1989.Fixture.SecondLevelCacheWithMixedCacheableAndNonCacheableFuture
NHibernate.Test.NHSpecificTest.NH1989.Fixture.SecondLevelCacheWithMixedCacheRegionsFuture
NHibernate.Test.NHSpecificTest.NH1989.Fixture.SecondLevelCacheWithSingleCacheableFuture
NHibernate.Test.NHSpecificTest.NH1989.Fixture.SecondLevelCacheWithSingleCacheableQueryFuture
NHibernate.Test.NHSpecificTest.NH2037.Fixture.Test
NHibernate.Test.NHSpecificTest.NH3221.WeirdBehaviour.CanAddATodo
NHibernate.Test.NHSpecificTest.NH3221.WeirdBehaviour.CanAddStuff
NHibernate.Test.NHSpecificTest.NH3221.WeirdBehaviour.CanRemoveATodo
NHibernate.Test.NHSpecificTest.NH3221.WeirdBehaviour.CanRemoveStuff

**Comment by lesnyh** [comment]
I think it may be a little bit hard to write a test, because now all NHibernate test dialects support not null unique constraints. I have found interesting comment in file NHibernate\Mapping\Index.cs:
//TODO handle supportsNotNullUnique=false, but such a case does not exist in the wild so far

File: src/NHibernate/Mapping/Table.cs [hunk 1]
```diff
@@ -414,9 +414,12 @@ public string SqlCreateString(Dialect.Dialect dialect, IMapping p, string defaul
 				buf.Append(StringHelper.CommaSpace).Append(PrimaryKey.SqlConstraintString(dialect, defaultSchema));
 			}
 
-			foreach (UniqueKey uk in UniqueKeyIterator)
+			if (dialect.SupportsNotNullUnique)
```
Comment (hazzik) [inline src/NHibernate/Mapping/Table.cs]: I think it's more accurate way is to check that `uk.SqlConstraintString(dialect)` returns not empty string before adding to `buf`

**Comment by lesnyh** [comment]
Can I add our LinterDialect class for DBMS Linter SQL ([https://linter.ru/en/](url))?

**Comment by gliljas** [comment]
That would be unnecessary and make the test bloated. Create a small dialect just for the test instead.,

**Comment by lesnyh** [comment]
I have corrected code and added small test.

File: src/NHibernate/Mapping/Table.cs (hunk 1 above)
Comment (lesnyh) [inline src/NHibernate/Mapping/Table.cs]: I have corrected code.

**Comment by lesnyh** [comment]
BTW, the property name SupportsNotNullUnique is a bit confusing because according to the usage of this property it means: "supports nullable columns with unique constraints". You'd better add a comment to clarify what this property mean.

**Comment by hazzik** [comment]
Can you please enable [Allow edits from maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) option on this PR?

**Comment by hazzik** [comment]
Replaced by #1651

**Comment by hazzik** [comment]
Duplicate of #1651

**Pull request closed without merging**